server/src/api-v1/communication/communication.gateway.ts

```diff
@@ -17,14 +17,14 @@ export class CommunicationGateway implements OnGatewayConnection, OnGatewayDisco
   ) {}
 
   @WebSocketServer() server: Server;
-
   handleDisconnect(client: Socket) {
     client.leave(client.nsp.name);
 
     Logger.log(`Client ${client.id} disconnected`);
   }
 
   async handleConnection(client: Socket, ...args: any[]) {
+    // todo handle websocket connection with oauth2
     Logger.log(`New websocket connection: ${client.id}`, 'NewWebSocketConnection');
     const accessToken = client.handshake.headers.authorization.split(' ')[1];
```

Comment:
I have a question regarding the handling of websockets: where is the WS connection used?

I'm asking because for what I've read online it's not possible to apply AuthGuards to the handleConnection() but only on the functions that handle specific messages. Since the verification of the JWT token is done by the already implemented strategies, it wouldn't be the nicest thing to rewrite the implementation of the token validation from scratch.
I think it would be better, in order to reduce code duplication and potential bugs, to find a way to use those strategies for the WebSockets too. What do you think?

Reply:
Websocket is used for communicating with the client when there is a new asset is uploaded to the server, so the client knows to fetch the latest uploaded assets without the user having to perform any refresh mechanism, like "pull to refresh"

Reply:
Oh okay thanks, I couldn't find anywhere in the code where this is used

Reply:
If you look into the mobile app folder, you can find the usage on the mobile app in `mobile/lib/shared/providers/websocket.provider.dart`

Reply:
Hi, I'd like your opinion on the implementation of the authentication for the WebSocket.

As of today (and it's been like that for 3 years, see [this](https://github.com/nestjs/nest/issues/882), NestJS doesn't allow to add AuthGuards to the handling of the connection.

So there is the (in my opinion) ugly solution which consists of reimplementing the same logic of the current Strategies (JWT verification with the secret key, expiration check and audience validation) and using that for the authentication during the initial phase of the websocket. That wouldn't be clean simply because of the code duplication.

The less ugly idea I found on the internet is to create an API calls that generates a ticket to be used only for one WS connection, during the `handleConnection()` the server verifies the presence of that token in the auth bearer and let the connection live if valid. This way the real authentication is handled by the HTTP call as all the other API calls, and the code would have way less bugs.

What do you think about this?

Reply:
Hey Enrico, sorry I missed this message earlier. Let me think about this.

Reply:
The second option sounds pretty solid to me, can have tokens specifically for the websockets generated by the client. How would it handle expiration of these tokens? Would it be a standard access token and refresh token combo, or one long-lived token?